proof/refine/ARM/Schedule_R.thy

```diff
@@ -3040,5 +2990,662 @@ lemma ct_released_cross_weak:
+              intro!: readTCBRefillReady_simp valid_release_q_active_sc)
+
+lemma ksReleaseQueue_length_well_founded:
+  "wf {((r :: unit, s :: kernel_state), (r', s')). length (ksReleaseQueue s)
+                                                    < length (ksReleaseQueue s')}"
+  apply (insert wf_inv_image[where r="{(m, n). m < n}"
+                               and f="\<lambda>(r :: unit, s :: kernel_state). length (ksReleaseQueue s)"])
+  apply (clarsimp simp: inv_image_def)
+  apply (prop_tac "wf {(m, n). m < n}")
+   apply (fastforce intro: wf)
+  apply (drule meta_mp)
+   apply simp
+  apply (prop_tac "{(x :: unit \<times> global.kernel_state, y ::  unit \<times> global.kernel_state).
+                     (case x of (r, s) \<Rightarrow> length (ksReleaseQueue s))
+                      < (case y of (r, s) \<Rightarrow> length (ksReleaseQueue s))}
+                   = {((r, s), r', s'). length (ksReleaseQueue s) < length (ksReleaseQueue s')}")
+   apply fastforce
+  apply fastforce
+  done
+
+lemma tcbReleaseDequeue_ksReleaseQueue_length_helper:
+  "\<lbrace>\<lambda>s'. ksReleaseQueue s' \<noteq> [] \<and> s' = s\<rbrace>
+   tcbReleaseDequeue
+   \<lbrace>\<lambda>r' s'. length (ksReleaseQueue s') < length (ksReleaseQueue s)\<rbrace>"
+  apply (wpsimp simp: tcbReleaseDequeue_def getReleaseQueue_def setReleaseQueue_def)
+  done
+
+lemma tcbReleaseDequeue_ksReleaseQueue_length:
+  "\<lbrace>\<lambda>s'. the (releaseQNonEmptyAndReady s') \<and> s' = s\<rbrace>
+   tcbReleaseDequeue
+   \<lbrace>\<lambda>r' s'. length (ksReleaseQueue s') < length (ksReleaseQueue s)\<rbrace>"
+  apply (wpsimp wp: tcbReleaseDequeue_ksReleaseQueue_length_helper)
+  apply (fastforce dest: releaseQNonEmptyAndReady_implies_releaseQNonEmpty)
+  done
+
+lemma awakenBody_release_queue_length:
+  "\<lbrace>\<lambda>s'. the (releaseQNonEmptyAndReady s') \<and> s' = s\<rbrace>
+   awakenBody
+   \<lbrace>\<lambda>r' s'. length (ksReleaseQueue s') < length (ksReleaseQueue s)\<rbrace>"
+  apply (clarsimp simp: awakenBody_def)
+  apply (wpsimp wp: tcbReleaseDequeue_ksReleaseQueue_length hoare_drop_imps)
+  done
+
+lemma awaken_terminates:
+  "whileLoop_terminates (\<lambda>_ s'. (the (releaseQNonEmptyAndReady s'))) (\<lambda>_. awakenBody) r' s'"
+  apply (rule_tac R="{((r', s'), (r, s)). length (ksReleaseQueue s') < length (ksReleaseQueue s)}"
+               in whileLoop_terminates_inv)
+    apply simp
+   apply clarsimp
+   apply (rule awakenBody_release_queue_length)
+  apply (rule ksReleaseQueue_length_well_founded)
+  done
+
+lemma reprogram_timer_update_release_queue_update_monad_commute:
+  "monad_commute \<top> (modify (reprogram_timer_update (\<lambda>_. b))) (modify (release_queue_update q))"
+  apply (clarsimp simp: monad_commute_def modify_def get_def put_def bind_def return_def)
+  done
+
+(* FIXME RT: move?*)
+lemma bind_return_noop:
+  "do y \<leftarrow> a; b od = do y \<leftarrow> a; z \<leftarrow> b; return () od"
+  by simp
```

Comment:
Is this still used? I am not sure if this is useful to keep if it is not used any more.

Reply:
Ah good catch. This wasn't used any more. I've removed it